### DISCUSSION TITLE
Unable to Replace new .exe with old .exe when i serve my app to windows Service Manager.

### DISCUSSION THREAD

**Discussion by ritishbhardwaj** [dennisvang/tufup#140] [Q&A] [answered]
Hi,
So basically, after implementing the functionality of autoUpdation, i have created the .bat file which is referring to the .exe [old one] file. Now with the following commands of nssm that i have saved in some .bat file, i have created the windows service.

```bat
@echo off
cd /d "%~dp0AutostartService"
:: Install the service
nssm.exe install demoApp1_app "%cd%\nssm_run.bat"
:: Set stdout and stderr logs
nssm.exe set demoApp1_app AppStdout "%cd%\logs\demoApp1_app_logs.log"
nssm.exe set demoApp1_app AppStderr "%cd%\logs\demoApp1_app_logs.log"
:: Configure log rotation
nssm set demoApp1_app AppRotateFiles 1
nssm set demoApp1_app AppRotateOnline 1
nssm set demoApp1_app AppRotateSeconds 86400
nssm set demoApp1_app AppRotateBytes 1048576
:: Start the service
sc start demoApp1_app
```
following is the structure of my .bat file
```bat
:loop
tasklist /nh /fi "imagename eq demoApp1.exe" | find /i "demoApp1.exe" > nul || (START "" /D "D:\AutoUpdationAPP1_UsingTUF\extra" "D:\AutoUpdationAPP1_UsingTUF\extra\demoApp1.exe")
timeout /t 17
goto loop`
```
Now this service is making the background process of the demoApp1.exe and i can recognise them in task manager
![Task_demoApp](https://github.com/dennisvang/tufup/assets/81488252/635979d7-1ae4-4690-9e04-df3d85df163b)

main Question : when i start updating my app [when app is appeared on foreground] and everything is downloaded , after that due to the process is being running in the background i am getting the following error....
`2024/05/14 05:53:27 ERROR 32 (0x00000020) Copying File C:\Users\ritis\AppData\Local\Temp\tufup\demoApp1.exe
The process cannot access the file because it is being used by another process.
Waiting 2 seconds... Retrying...`

Can you guide me that is there any way through which i can end or remove all the tasks running in the processes for this particular app name using the TUFUP when it downloads everything from the server, so that smooth updation can be completed, as the .bat file will run the app automatically once it get closed.
[ If i am wrong ,kindly correct me. And am i asking the things right or do i need to make changes in my .bat file code.]

following is my TUFUP Client code.
```python
def updateHandler(install_dir:str,meta_dir:str,target_dir:str,update:bool):
    client = tufup.client.Client(APP_NAME, install_dir, #it is app install directory on client machine 
                                 __version__, meta_dir, #this is meta directory   on client machine 
                                 METADATA_BASE_URL,
                                 target_dir , TARGET_BASE_URL)

    # client.refresh()
    any_update = client.check_for_updates()
    if update ==True:
        return any_update
    
    print("AnyUpdate---------->    ",any_update)
    # getInfo = client.get_targetinfo(DEV_DIR / 'repository/metadata')
    # print(getInfo)
    if any_update:
        print("woooooohoooooooo")

        if any_update.custom:
            print('changes in this update:')
            for item in any_update.custom.get('changes', []):
                print(f'\t- {item}')

        # path_to_updateBat = 'D:\\AutoUpdationAPP1_UsingTUF\\AutostartService\\update_bat.bat'
        # subprocess.run(path_to_updateBat,shell=True)
        
        # client.download_and_apply_update(progress_hook=progress_hook,skip_confirmation=True,process_creation_flags=subprocess.CREATE_NO_WINDOW)
        client.download_and_apply_update(progress_hook=progress_hook,skip_confirmation=True)
        print("Downloading......")
```

Waiting for your reply...... :)

**Comment by dennisvang** [comment]
Hi @ritishbhardwaj, if I understand correctly, your service continually looks for your app processes, and starts a new process if it does not find one. Is that correct?

Your screenshot shows two processes for demoApp1.exe. Do you need multiple or do you normally have only one process running?

**Comment by dennisvang** [comment, marked as answer]
In any case, keep in mind that tufup "installs" an update by overwriting the content of your app's installation directory.

On windows, this is not possible as long as the app itself is still running, because the executable file is still in use. For more information, see e.g. https://stackoverflow.com/q/196897.

For this reason, when the tufup client "installs" an update, it will first start a separate (independent) process that runs the windows `robocopy` utility ([source][1]), and then exits the running app's process ([source][2]). As long as the app process has not exited completely, robocopy will fail, and another attempt is made after a short timeout. This is what you are seeing in the error:

>... The process cannot access the file because it is being used by another process. Waiting 2 seconds... Retrying...

If your app is running in multiple processes, you will need to terminate all those processes, otherwise robocopy will keep failing.

[1]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/utils/platform_specific.py#L88
[2]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/utils/platform_specific.py#L233

**Reply by ritishbhardwaj** [reply]
Hi Thanks for this. I have found some other way before my application gets updated.
:)

**Reply by ritishbhardwaj** [reply]
@dennisvang  New problem is getting up, suppose i have placed the .exe in the c:\Program files (X86)\myApp.
Now when i am updating the app in this folder, an error comes like unable to replace the app in this folder, but when i run the app as an administrator then it is updating the app gracefully. Is there any way to deal with it within the tufup, like Client code will ask the user to make sure he is in admin mode before installing the application, or is there any other way.

[ Also when i copy the same old .exe in d:\ folder , the app is getting updated even without opening as admin, [ so it gives me hint that there is an issue of admin rights ].

Waiting for your reply !!!!!!!!!!!

**Reply by dennisvang** [reply]
> ... Is there any way to deal with it within the tufup, like Client code will ask the user to make sure he is in admin mode before installing the application, or is there any other way.

Hi @ritishbhardwaj, yes, there is another way. On windows you can automatically request elevation by setting `as_admin=True`, as follows ([source][1]):

```python
client.download_and_apply_update(..., as_admin=True)
```

As a result, windows will show the [UAC elevation prompt][5], allowing you to enter the admin password.

For those interested, this is implemented in [`tufup.utils.platform_specific.run_bat_as_admin()`][6].

Alternatively, you could do a per-user installation, typically in the `%LOCALAPPDATA%\Programs` folder, which does not require admin rights. See e.g. [Windows installation context docs][3], more about [known folders][4], and [this post][2].

[1]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/utils/platform_specific.py#L131
[2]: https://superuser.com/q/1482669
[3]: https://learn.microsoft.com/en-us/windows/win32/msi/installation-context
[4]: https://learn.microsoft.com/en-us/windows/win32/shell/knownfolderid
[5]: https://learn.microsoft.com/en-us/windows/security/application-security/application-control/user-account-control/how-it-works
[6]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/utils/platform_specific.py#L99

**Reply by ritishbhardwaj** [reply]
@dennisvang , Hi Dennisvang
I have tried it and it is successfully updating my app after asking me for the admin permissions. Thanks for your help !!!!!

**But** before your above reply what i think to try ,is to install a ```batch``` file with nssm , which runs as a windows service [i.e Batch file will now be run as service]  which checks in infinite loop whether there is new file in the particular directory in the  ```%LOCALAPPDATA%\Programs```  and if it detects a new file overthere, ( the new file will come/downloaded after i press the button to download, on my app and it just do download operation and no updation) , it will write the new file in the ```c:\ProgramFiles``` without asking me to enter admin rights because a service, according to my research won't need admin rights.  [ Will tell you result , after i make a try (even i don't know whether this could be possible or not)  ]
Your Solution works for me fine but just out of curiosity , can i know  whether we can do the same [like not even asking for admin details] using some parameter directly into the function. :)

Also could you tell me, from where you are referring all these things -----> ```client.download_and_apply_update(..., as_admin=True)``` , i want to explore by my own, and also there is one parameter in ```download_and_apply_update(...)``` named as ```install``` which is of callable type, can you tell me what is the context behind it, like when i see the code , it is bit confusing.
One more thing, in the same function ```download_and_apply_update(...)``` we have **kwargs i.e. we can pass keyword arguments to it, but  can you tell me which key-word arguments will it accept and what valid arguments.....

Waiting for your reply for this too :)

**Reply by dennisvang** [reply]
>... But before your above reply what i think to try ,is to install a batch file with nssm , which runs as a windows service [...] which checks in infinite loop whether there is new file in the particular directory in the %LOCALAPPDATA%\Programs and if it detects a new file overthere, [...], it will write the new file in the c:\ProgramFiles without asking me to enter admin rights because a service, according to my research won't need admin rights. ...

@ritishbhardwaj As I don't have any experience with services, I'm not sure if that would work. I do expect at some point you would need to have admin rights to allow writing anything to `%ProgramFiles%`, e.g. when installing the service.

Your approach does sound quite complex. A simple per-user install would be much easier, unless you really require per-machine installation for some reason.

>Your Solution works for me fine but just out of curiosity , can i know whether we can do the same [like not even asking for admin details] using some parameter directly into the function. :)

Not really, except for specifying a different install dir. The install dir determines whether admin rights are required. 

>Also could you tell me, from where you are referring all these things ...

See all the links in my previous reply.

>... there is one parameter in download_and_apply_update(...) named as install which is of callable type, can you tell me what is the context behind it, like when i see the code , it is bit confusing.

The `install` argument accepts a custom installation function. You can use it to do whatever you want in the installation step. If you specify a custom `install` function, that will be used instead of [`_install_update_win()`][1].

>One more thing, in the same function download_and_apply_update(...) we have **kwargs i.e. we can pass keyword arguments to it, but can you tell me which key-word arguments will it accept and what valid arguments.....

Any additional `kwargs` are passed on to the `install` callable. By [default][2], on Windows, that's [`_install_update_win()`][1]. Also see [`Client._apply_update()`][3].

This seems a bit complicated, but it is necessary for flexibility, in order to allow customization.

[1]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/utils/platform_specific.py#L126
[2]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/client.py#L131
[3]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/client.py#L325

**Comment by ritishbhardwaj** [comment]
hi @dennisvang  any reply for the above [https://github.com/dennisvang/tufup/discussions/140#discussioncomment-9546344] ?
